### PR TITLE
Fix OVTrainer IR saving for structurally pruned model

### PR DESCRIPTION
**Pull request by yujiepan-work**
## Background
To save the structurally pruned IR model, the current workaround is:  
dynamic shape onnx -> dynamic shape IR ->  static shape IR -> pruned static shape IR -> pruned dynamic shape IR.

Swin model will get errors at "dynamic shape IR ->  static shape IR " step because swin cannot accept arbitrary static shape. To fix this, this PR proposes to use the input info in `ovconfig` to get a proper static shape.

## Changes
- "dynamic shape IR ->  static shape IR " step in generating pruned IR will now use `ovconfig`'s input info as static shape.
- If the above step failed, then IR will not be saved, so that users can realize that something wrong happens.
- Add new tests to check ovtrainer can save swin / wav2vec models with structured pruning configs. 
- The new tests are in `test_trainer.py`.  Currently, the file contains only the ovtrainer tests on **text-classification**.  This PR extends new cases on image-classification and audio-classification because we want to check swin and wav2vec2 models.  To better organize these 3 cases, we abstract a new **base class** to aggregate common codes, and create the test classes specific for each case. Though the codes are highly refactored, existing checks on text-classification are 100%  preserved.
- Add a new test requirement `librosa` for audio task tests, see [license](https://github.com/librosa/librosa/blob/main/LICENSE.md)

### CODE REVIEW

**Comment by yujiepan-work** [comment]
Hi, @AlexKoff88 @helena-intel, this PR fixes the IR saving for swin model when it is structurally pruned by JPQD. Could you help to look at it?

**Review by AlexKoff88 — APPROVED** [review]

**Comment by vuiseng9** [comment]
This PR should be merged before https://github.com/huggingface/optimum-intel/pull/210

File: optimum/intel/openvino/trainer.py [hunk 1]
```diff
@@ -719,8 +719,10 @@ def _save(self, output_dir: Optional[str] = None, state_dict=None):
                     if not save_as_external_data:
                         onnx_path.write_bytes(f.getvalue())
                     logger.warning(
-                        f"Error encountered during IR pruning: {err}. {onnx_path} is dumped for debug. Run continues."
+                        f"Error encountered during OpenVINO IR pruning: {err}. {onnx_path} is dumped for debugging. "
```
Comment (bosbes) [inline optimum/intel/openvino/trainer.py:722]: I expect most users will not know what to do with the ONNX model for debugging. It's a blanket exception, it's a bit risky to only show a warning. Shouldn't we reraise the exception and then still dump the ONNX model? Then it's at least clear that something is broken, and you get the exact error, instead of just a generic warning that may be easy to miss.

File: optimum/intel/openvino/trainer.py [hunk 2]
```diff
@@ -749,9 +751,12 @@ def _should_apply_pruning_transform(self) -> bool:
 
     def _reshape_ir(self, ov_model: openvino.runtime.Model, static_shape: bool) -> openvino.runtime.Model:
         new_input_cfg = dict()
+        input_name_vs_shape = {item["keyword"]: item["sample_size"] for item in self.ov_config.input_info}
```
Comment (bosbes) [inline optimum/intel/openvino/trainer.py:754]: Is there a possibility that `self.ov_config.input_info` is not defined at this point?

File: optimum/intel/openvino/trainer.py [hunk 3]
```diff
@@ -719,8 +719,10 @@ def _save(self, output_dir: Optional[str] = None, state_dict=None):
                     if not save_as_external_data:
                         onnx_path.write_bytes(f.getvalue())
                     logger.warning(
-                        f"Error encountered during IR pruning: {err}. {onnx_path} is dumped for debug. Run continues."
+                        f"Error encountered during OpenVINO IR pruning: {err}. {onnx_path} is dumped for debugging. "
+                        "Run continues without saving OpenVINO IR files."
                     )
+                    return
             else:
                 if self._get_compression_controller_by_cls(QuantizationController) is not None:
```
Comment (helena-intel) [inline optimum/intel/openvino/trainer.py:727]: It's a blanket exception, it's a bit risky to only show a warning. Shouldn't we reraise the exception and then still dump the ONNX model? Then it's clear that something is broken, and you get the exact error

File: optimum/intel/openvino/trainer.py (hunk 2 above)
Comment (helena-intel) [inline optimum/intel/openvino/trainer.py:754]: Is there a possibility that self.ov_config.input_info is not defined at this point?

File: optimum/intel/openvino/trainer.py (hunk 3 above)
Comment (yujiepan-work) [reply optimum/intel/openvino/trainer.py:726]: > It's a blanket exception, it's a bit risky to only show a warning. Shouldn't we reraise the exception and then still dump the ONNX model? Then it's clear that something is broken, and you get the exact error

I agree with your concern. See updates: https://github.com/huggingface/optimum-intel/blob/5618c3cf45d61bab48d702a0254001395e66f182/optimum/intel/openvino/trainer.py#L717-L724

File: optimum/intel/openvino/trainer.py (hunk 2 above)
Comment (yujiepan-work) [reply optimum/intel/openvino/trainer.py:753]: Line176, Lind196

File: tests/openvino/test_training.py [hunk 4]
```diff
@@ -116,317 +98,737 @@ class OVTrainerTestDescriptor:
...
+                for inputs in self.trainer.get_eval_dataloader(dataset):
+                    self.assertSequenceEqual(inputs["input_ids"].shape, [batch_size, seq_length])
+                    ovmodel_outputs = ovmodel(**inputs)
+                    self.assertIn("logits", ovmodel_outputs)
+                    ovmodel_logits = ovmodel_outputs.logits
+                    with torch.no_grad():
+                        torch_logits = torch_model(**inputs).logits
+                    torch.testing.assert_close(
+                        torch.softmax(ovmodel_logits, dim=-1),
+                        torch.softmax(torch_logits, dim=-1),
```
Comment (echarlaix) [inline tests/openvino/test_training.py:500]: Could this be replaced with :
```suggestion
                        ovmodel_logits,
                        torch_logits,
```

**Review by echarlaix — APPROVED** [review]
It looks great, thanks a lot @yujiepan-work !

File: tests/openvino/test_training.py (hunk 4 above)
Comment (yujiepan-work) [reply tests/openvino/test_training.py]: The softmax here is intended to amplify the logits so that atol=1e-4 can pass.

I check the model logits and find that sometimes the values are very close to zero, e.g., 0.0071. In **quantization** cases where higher error is expected than fp32 models, the original logit values are easy to accidentally fail at atol=1e-4 (e.g., 0.0071 vs. 0.0073).

In the latest commit I have deleted the softmax, but I also have to increase atol to 1e-3 to pass the tests (see latest checks here， ignoring the INC part which should not our problem). Is that acceptable? I see some other tests are also using 1e-3, e.g.,
https://github.com/huggingface/optimum-intel/blob/844be9adc9eb3bea8f87c22fa9112a4fb3a31d88/tests/openvino/test_modeling.py#L646

Comment (echarlaix) [reply tests/openvino/test_training.py]: A tolerance of 1e-3 sounds reasonable, thanks a lot for updating it !
